### ISSUE TITLE
Discrepancy in ROTATION_TRICK Output: emb vs emb_out Mismatch

### ISSUE THREAD

**Issue by skyler-19** [EdoardoBotta/RQ-VAE-Recommender#54]
When using ROTATION_TRICK, the values of emb and emb_out are not identical, which contradicts the corresponding formula.

code: 
if self.training:
    if self.forward_mode == QuantizeForwardMode.GUMBEL_SOFTMAX:
        weights = gumbel_softmax_sample(
            -dist, temperature=temperature, device=self.device
        )
        emb = weights @ codebook
        emb_out = emb
    elif self.forward_mode == QuantizeForwardMode.STE:
        emb = self.get_item_embeddings(ids)
        emb_out = x + (emb - x).detach()

    elif self.forward_mode == QuantizeForwardMode.ROTATION_TRICK:
        emb = self.get_item_embeddings(ids)
        emb_out = efficient_rotation_trick_transform(
            x / (x.norm(dim=-1, keepdim=True) + 1e-8),
            emb / (emb.norm(dim=-1, keepdim=True) + 1e-8),
            x
        )
    else:
        raise Exception("Unsupported Quantize forward mode.")
    
    loss = self.quantize_loss(query=x, value=emb)
else:
    emb_out = self.get_item_embeddings(ids)
    loss = self.quantize_loss(query=x, value=emb_out)

#debug
diff = ((emb - emb_out)**2).sum(dim=-1).mean().item()
print(diff)

The diff reaches values around 0.2 to 0.3 . Additionally, when I set up multi-layer quantization , the error increases significantly.

**Comment by skyler-19** [comment]
Should the Correct Implementation Scale emb_out by Norm Ratio


```
    emb = self.get_item_embeddings(ids)
    emb_out = efficient_rotation_trick_transform(
        x / (x.norm(dim=-1, keepdim=True) + 1e-8),
        emb / (emb.norm(dim=-1, keepdim=True) + 1e-8),
        x
    )
    emb_out = emb_out * (
        torch.norm(emb, dim=1, keepdim=True) / (torch.norm(x, dim=1, keepdim=True) + 1e-6)
    ).detach()
```

**Comment by EdoardoBotta** [comment]
you're right, I have updated the code.
